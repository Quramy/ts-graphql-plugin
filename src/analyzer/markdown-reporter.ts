import path from 'path';
import { ManifestOutput, ManifestDocumentEntry } from './types';

export type ToMarkdownContentOptions = {
  baseDir: string;
  outputDir: string;
  ignoreFragments?: boolean;
};

function createLinkPath(fileName: string, baseDir: string, outputDir: string) {
  const displayPath = path.isAbsolute(fileName) ? path.relative(baseDir, fileName) : fileName;
  const linkPath = path.isAbsolute(fileName) ? path.relative(outputDir, fileName) : fileName;
  return { displayPath, linkPath };
}

function createSection(
  sectionName: string,
  operationDocs: ManifestDocumentEntry[],
  baseDir: string,
  outputDir: string,
) {
  const h2 = '## ' + sectionName[0].toUpperCase() + sectionName.slice(1);
  return (
    h2 +
    '\n' +
    operationDocs
      .map(doc => {
        const { displayPath, linkPath } = createLinkPath(doc.fileName, baseDir, outputDir);
        return `
### ${doc.type !== 'fragment' ? doc.operationName || 'anonymous' : doc.fragmentName}

\`\`\`graphql
${doc.body.trim()}
\`\`\`

From [${displayPath}:${doc.documentStart.line + 1}:${doc.documentStart.character + 1}](${linkPath}#L${
          doc.documentStart.line + 1
        }-L${doc.documentEnd.line + 1})
    `;
      })
      .join('\n')
  );
}

export class MarkdownReporter {
  toMarkdownConntent(
    manifest: ManifestOutput,
    { ignoreFragments = true, baseDir, outputDir }: ToMarkdownContentOptions,
  ) {
    if (!manifest.documents.length) return null;
    const outs = ['# Extracted GraphQL Operations'];
    const groupedDocs = {
      queries: [],
      mutations: [],
      subscriptions: [],
      fragments: [],
    } as { [key: string]: ManifestDocumentEntry[] };
    manifest.documents.forEach(doc => {
      switch (doc.type) {
        case 'query':
          groupedDocs.queries.push(doc);
          break;
        case 'mutation':
          groupedDocs.mutations.push(doc);
          break;
        case 'subscription':
          groupedDocs.subscriptions.push(doc);
          break;
        case 'fragment':
          if (!ignoreFragments) groupedDocs.fragments.push(doc);
          break;
        default:
          break;
      }
    });
    Object.entries(groupedDocs).forEach(([name, docs]) => {
      if (docs.length) {
        outs.push(createSection(name, docs, baseDir, outputDir));
      }
    });
    outs.push('---');
    outs.push('Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)');
    return outs.join('\n');
  }
}
