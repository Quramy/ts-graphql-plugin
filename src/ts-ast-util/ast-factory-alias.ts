import ts from 'typescript';

export const astf = ts.factory;
