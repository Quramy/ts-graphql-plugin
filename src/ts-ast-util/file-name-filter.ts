import _path from 'node:path';

import { globToRegExp } from '../string-util/glob-to-regexp';

export function createFileNameFilter({
  specs,
  projectName,
  _forceWin32 = false,
}: {
  specs: string[] | undefined;
  projectName: string;
  _forceWin32?: boolean;
}) {
  const path = _forceWin32 ? _path.win32 : _path;

  const testers = (specs ?? []).map(pattern => {
    if (pattern.includes('*') || pattern.includes('?')) {
      const regexp = globToRegExp(pattern);
      return (normalized: string) => regexp.test(normalized);
    } else {
      const dirOrFileName = pattern[pattern.length - 1] === '/' ? pattern.slice(0, pattern.length - 1) : pattern;
      return (normalized: string) => normalized === dirOrFileName || normalized.startsWith(dirOrFileName + '/');
    }
  });

  const projectRootDirName = path.dirname(projectName);

  const match = (fileName: string) => {
    const normalized = path.relative(projectRootDirName, fileName).replace(/\\/g, '/');
    return testers.some(tester => tester(normalized));
  };

  return match;
}
