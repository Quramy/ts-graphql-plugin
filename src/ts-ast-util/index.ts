import ts from 'typescript/lib/tsserverlibrary';

export * from './types';

export * from './template-expression-resolver';

export { createScriptSourceHelper } from './script-source-helper';

export type TagCondition = string;

export function findNode(sourceFile: ts.SourceFile, position: number): ts.Node | undefined {
  function find(node: ts.Node): ts.Node | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
  }
  return find(sourceFile);
}

export function findAllNodes(sourceFile: ts.SourceFile, cond: (n: ts.Node) => boolean): ts.Node[] {
  const result: ts.Node[] = [];
  function find(node: ts.Node) {
    if (cond(node)) {
      result.push(node);
      return;
    } else {
      ts.forEachChild(node, find);
    }
  }
  find(sourceFile);
  return result;
}

export function isTagged(node: ts.Node, condition: TagCondition) {
  if (!node || !node.parent) return false;
  if (!ts.isTaggedTemplateExpression(node.parent)) return false;
  const tagNode = node.parent;
  return tagNode.tag.getText() === condition;
}

export function isTemplateLiteralTypeNode(node: ts.Node): node is ts.TemplateLiteralTypeNode {
  // ts.isNoSubstitutionTemplateLiteral exists TypeScript >= 4.1
  return typeof ts.isTemplateLiteralTypeNode === 'function' && ts.isTemplateLiteralTypeNode(node);
}
