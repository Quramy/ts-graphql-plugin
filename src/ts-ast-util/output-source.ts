import path from 'path';
import ts from 'typescript';

import { OutputSource } from './types';
import {
  createImportSpecifier,
  isImportDeclarationWithCondition,
  mergeImportDeclarationsWithSameModules,
} from './utilily-functions';

const printer = ts.createPrinter();

/**
 *
 * @internal
 *
 */
export class DefaultOutputSource implements OutputSource {
  private _unmergedComment: string[] = [];
  private _statements: ts.Statement[] = [];

  constructor(private _filename: string) {}

  get outputFileName() {
    return this._filename;
  }

  get outputDirName() {
    return path.dirname(this._filename);
  }

  findImportDeclarationIndex({ isDefault, name, from }: { isDefault?: boolean; name?: string; from?: string }) {
    return this._statements.findIndex(st => isImportDeclarationWithCondition(st, { isDefault, name, from }));
  }

  pushNamedImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from }) !== -1) return false;
    const importSpecifier = createImportSpecifier(false, undefined, ts.createIdentifier(identifierName));
    const namedBinding = ts.createNamedImports([importSpecifier]);
    const importClause = ts.createImportClause(undefined, namedBinding);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushDefaultImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from, isDefault: true }) !== -1) return false;
    const importClause = ts.createImportClause(ts.createIdentifier(identifierName), undefined);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushImportDeclaration(declaration: ts.ImportDeclaration) {
    for (let i = 0; i < this._statements.length; i++) {
      if (!ts.isImportDeclaration(this._statements[i])) {
        this._insertStatement(i, declaration);
        return;
      }
    }
    this._insertStatement(0, declaration);
  }

  pushStatement(statement: ts.Statement) {
    this._insertStatement(this._statements.length, statement);
  }

  replaceStatement(statement: ts.Statement, newStatement: ts.Statement) {
    const idx = this._findIndex(statement);
    if (idx === -1) return false;
    this._replaceStatement(idx, newStatement);
    return true;
  }

  removeStatement(statement: ts.Statement) {
    const idx = this._findIndex(statement);
    if (idx === -1) return false;
    this._statements = [...this._statements.slice(0, idx), ...this._statements.slice(idx + 1)];
    return true;
  }

  writeLeadingComment(comment: string) {
    if (this._statements.length > 0) {
      ts.addSyntheticLeadingComment(this._statements[0], ts.SyntaxKind.MultiLineCommentTrivia, ` ${comment} `, true);
    } else {
      this._unmergedComment.push(comment);
    }
  }

  getStatements() {
    return this._statements as ReadonlyArray<ts.Statement>;
  }

  toSourceFile() {
    const sourceFile = ts.createSourceFile(this._filename, '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
    return ts.updateSourceFileNode(sourceFile, this._statements);
  }

  toFileContent() {
    const content = printer.printFile(this.toSourceFile());
    return { fileName: this._filename, content };
  }

  private _findIndex(statement: ts.Statement) {
    return this._statements.findIndex(st => st === statement);
  }

  private _insertStatement(index: number, statement: ts.Statement) {
    this._statements = [...this._statements.slice(0, index), statement, ...this._statements.slice(index)];
    if (this._unmergedComment.length) {
      this._unmergedComment.forEach(comment => this.writeLeadingComment(comment));
      this._unmergedComment = [];
    }
  }

  private _replaceStatement(index: number, statement: ts.Statement) {
    if (!this._statements[index]) return;
    this._statements[index] = statement;
  }
}

export function createOutputSource({ outputFileName }: { outputFileName: string }) {
  return new DefaultOutputSource(outputFileName);
}
