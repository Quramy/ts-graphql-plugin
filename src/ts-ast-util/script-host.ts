import ts from 'typescript';

export class ScriptHost implements ts.LanguageServiceHost {
  private readonly _fileMap = new Map<string, string | undefined>();
  private readonly _fileVersionMap = new Map<string, number>();

  constructor(private readonly _currentDirectory: string, private readonly _compilerOptions: ts.CompilerOptions) {}

  readFile(fileName: string) {
    const hit = this._fileMap.get(fileName);
    if (hit != null) return hit;
    return this.loadFromFileSystem(fileName);
  }

  loadFromFileSystem(fileName: string) {
    const content = ts.sys.readFile(fileName, 'uts8');
    this._updateFile(fileName, content);
    return content;
  }

  getCurrentDirectory() {
    return this._currentDirectory;
  }

  getScriptSnapshot(fileName: string) {
    const file = this._fileMap.get(fileName);
    if (file == null) return;
    return ts.ScriptSnapshot.fromString(file);
  }

  getScriptVersion(fileName: string) {
    const version = this._fileVersionMap.get(fileName);
    if (!version) return '0';
    return version + '';
  }

  getScriptFileNames() {
    return [...this._fileMap.keys()];
  }

  getCompilationSettings() {
    return this._compilerOptions;
  }

  getDefaultLibFileName(opt: ts.CompilerOptions) {
    return ts.getDefaultLibFileName(opt);
  }

  // Sinse ts 4.7, imeplement required.
  // Delegating to `ts.sys.fileExists(path)` is not for me, but I don't know whether the following implementation is correct.
  fileExists() {
    return false;
  }

  protected _updateFile(fileName: string, content: string | undefined) {
    this._fileMap.set(fileName, content);
    const currentVersion = this._fileVersionMap.get(fileName) || 0;
    this._fileVersionMap.set(fileName, currentVersion + 1);
  }
}
