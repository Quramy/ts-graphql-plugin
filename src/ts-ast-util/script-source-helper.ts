import ts from 'typescript';
import { ScriptSourceHelper } from './types';
import { findAllNodes, findNode } from './utilily-functions';
import { TemplateExpressionResolver } from './template-expression-resolver';

export function createScriptSourceHelper({
  languageService,
  languageServiceHost,
}: {
  languageService: ts.LanguageService;
  languageServiceHost: ts.LanguageServiceHost;
}): ScriptSourceHelper {
  const getSourceFile = (fileName: string) => {
    const program = languageService.getProgram();
    if (!program) {
      throw new Error('language service host does not have program!');
    }
    const s = program.getSourceFile(fileName);
    if (!s) {
      throw new Error('No source file: ' + fileName);
    }
    return s;
  };
  const getNode = (fileName: string, position: number) => {
    return findNode(getSourceFile(fileName), position);
  };
  const getAllNodes = (fileName: string, cond: (n: ts.Node) => boolean) => {
    const s = getSourceFile(fileName);
    return findAllNodes(s, cond);
  };
  const getLineAndChar = (fileName: string, position: number) => {
    const s = getSourceFile(fileName);
    return ts.getLineAndCharacterOfPosition(s, position);
  };
  const resolver = new TemplateExpressionResolver(languageService, (fileName: string) =>
    languageServiceHost.getScriptVersion(fileName),
  );
  const resolveTemplateLiteral = resolver.resolve.bind(resolver);
  const updateTemplateLiteralInfo = resolver.update.bind(resolver);
  return {
    getNode,
    getAllNodes,
    getLineAndChar,
    resolveTemplateLiteral,
    updateTemplateLiteralInfo,
  };
}
