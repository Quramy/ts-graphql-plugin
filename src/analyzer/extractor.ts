import ts from 'typescript';
import { parse, print, DocumentNode, GraphQLError } from 'graphql';
import { visit } from 'graphql/language';
import { isTagged, ScriptSourceHelper, ResolvedTemplateInfo } from '../ts-ast-util';
import { ManifestOutput, ManifestDocumentEntry, OperationType } from './types';
import { ErrorWithLocation, ERROR_CODES } from '../errors';
import { detectDuplicatedFragments } from '../gql-ast-util';

export type ExtractorOptions = {
  removeDuplicatedFragments: boolean;
  scriptSourceHelper: ScriptSourceHelper;
  debug: (msg: string) => void;
};

export type ExtractTemplateResolveErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolveTemplateError: {
    message: string;
    start: number;
    end: number;
  };
  documentNode: undefined;
  resolevedTemplateInfo: undefined;
  graphqlError: undefined;
};

export type ExtractGraphQLErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: GraphQLError;
  resolveTemplateErrorMessage: undefined;
  documentNode: undefined;
};

export type ExtractSucceededResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  documentNode: DocumentNode;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: undefined;
  resolveTemplateErrorMessage: undefined;
};

export type ExtractResult = ExtractTemplateResolveErrorResult | ExtractGraphQLErrorResult | ExtractSucceededResult;

export class Extractor {
  private readonly _removeDuplicatedFragments: boolean;
  private readonly _helper: ScriptSourceHelper;
  private readonly _debug: (msg: string) => void;

  constructor({ debug, removeDuplicatedFragments, scriptSourceHelper }: ExtractorOptions) {
    this._removeDuplicatedFragments = removeDuplicatedFragments;
    this._helper = scriptSourceHelper;
    this._debug = debug;
  }

  extract(files: string[], tagName?: string): ExtractResult[] {
    const results: ExtractResult[] = [];
    this._debug('Extract template literals from: ');
    this._debug(files.map(f => ' ' + f).join(',\n'));
    files.forEach(fileName => {
      const nodes = this._helper
        .getAllNodes(fileName, node => ts.isTemplateExpression(node) || ts.isNoSubstitutionTemplateLiteral(node))
        .filter(node => (tagName ? isTagged(node, tagName) : true)) as (
        | ts.TemplateExpression
        | ts.NoSubstitutionTemplateLiteral
      )[];
      nodes.forEach(node => {
        const { resolvedInfo, resolveErrors } = this._helper.resolveTemplateLiteral(fileName, node);
        if (!resolvedInfo) {
          resolveErrors
            .filter(re => re.fileName === fileName)
            .forEach(resolveError => {
              results.push({
                fileName,
                templateNode: node,
                resolveTemplateError: {
                  message: ERROR_CODES.templateIsTooComplex.message,
                  start: resolveError.start,
                  end: resolveError.end,
                },
                resolevedTemplateInfo: undefined,
                graphqlError: undefined,
                documentNode: undefined,
              });
            });
        } else {
          results.push({
            fileName,
            templateNode: node,
            resolevedTemplateInfo: resolvedInfo,
          } as any);
        }
      });
    });
    return results.map(result => {
      if (!result.resolevedTemplateInfo) return result;
      try {
        const rawDocumentNode = parse(result.resolevedTemplateInfo.combinedText);
        if (!this._removeDuplicatedFragments) {
          return {
            ...result,
            documentNode: rawDocumentNode,
          } as ExtractSucceededResult;
        }
        const duplicatedInfo = detectDuplicatedFragments(rawDocumentNode);
        const updatedResolvedInfo = duplicatedInfo.reduce(
          (acc, fragmentInfo) => this._helper.updateTemplateLiteralInfo(acc, fragmentInfo),
          result.resolevedTemplateInfo,
        );
        const documentNode = parse(updatedResolvedInfo.combinedText);
        return {
          ...result,
          documentNode,
          resolevedTemplateInfo: updatedResolvedInfo,
        } as ExtractSucceededResult;
      } catch (error) {
        if (error instanceof GraphQLError) {
          return {
            ...result,
            graphqlError: error,
          } as ExtractGraphQLErrorResult;
        } else {
          throw error;
        }
      }
    });
  }

  pickupErrors(
    extractResults: ExtractResult[],
    { ignoreGraphQLError }: { ignoreGraphQLError: boolean } = { ignoreGraphQLError: false },
  ) {
    const errors: ErrorWithLocation[] = [];
    extractResults.forEach(r => {
      if (r.resolveTemplateError) {
        errors.push(
          new ErrorWithLocation(r.resolveTemplateError.message, {
            fileName: r.fileName,
            severity: 'Warn',
            content: r.templateNode.getSourceFile().getFullText(),
            start: r.resolveTemplateError.start,
            end: r.resolveTemplateError.end,
          }),
        );
      } else if (!ignoreGraphQLError && r.graphqlError) {
        const innerLoc = r.graphqlError.locations && r.graphqlError.locations[0];
        if (!innerLoc) {
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start: r.templateNode.getStart(),
              end: r.templateNode.getEnd(),
            }),
          );
        } else if (r.resolevedTemplateInfo) {
          const info = r.resolevedTemplateInfo;
          const innerPosition = info.convertInnerLocation2InnerPosition({
            line: innerLoc.line - 1,
            character: innerLoc.column - 1,
          });
          const start = info.getSourcePosition(innerPosition).pos;
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start,
              end: start + 1,
            }),
          );
        }
      }
    });
    return errors;
  }

  getDominantDefinition(result: ExtractSucceededResult) {
    let type: OperationType | undefined;
    const definedFragmentNames: string[] = [];
    const referencedFragmentNames: string[] = [];
    let operationName: string | undefined;
    visit(result.documentNode, {
      FragmentDefinition(node) {
        if (!type) {
          type = 'fragment';
        }
        definedFragmentNames.push(node.name.value);
      },
      FragmentSpread(node) {
        referencedFragmentNames.push(node.name.value);
      },
      OperationDefinition(node) {
        if (!type || type === 'fragment') {
          type = node.operation;
        } else {
          type = 'complex';
        }
        if (!operationName) {
          operationName = node.name ? node.name.value : 'ANONYMOUS_QUERY';
        } else {
          operationName = 'MULTIPLE_OPERATIONS';
        }
      },
    });
    const noReferedFragmentNames = definedFragmentNames.filter(defName =>
      referencedFragmentNames.every(n => defName !== n),
    );
    return { type, operationName, fragmentName: noReferedFragmentNames[noReferedFragmentNames.length - 1] };
  }

  toManifest(extractResults: ExtractResult[], tagName: string = ''): ManifestOutput {
    const documents = extractResults
      .filter(r => !!r.documentNode)
      .map(result => {
        const r = result as ExtractSucceededResult;
        const { type, operationName, fragmentName } = this.getDominantDefinition(r);
        return {
          fileName: r.fileName,
          type: type || 'other',
          operationName,
          fragmentName,
          body: print(r.documentNode!),
          tag: tagName,
          templateLiteralNodeStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart()),
          templateLiteralNodeEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd()),
          documentStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart() + 1),
          documentEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd() - 1),
        } as ManifestDocumentEntry;
      });
    return {
      documents,
    };
  }
}
