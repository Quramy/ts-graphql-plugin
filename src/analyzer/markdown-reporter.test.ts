import { Extractor } from './extractor';
import { MarkdownReporter } from './markdown-reporter';
import { createTestingLanguageServiceAndHost } from '../ts-ast-util/testing/testing-language-service';
import { createScriptSourceHelper } from '../ts-ast-util/script-source-helper';

function createExtractor(files: { fileName: string; content: string }[]) {
  const { languageService, languageServiceHost } = createTestingLanguageServiceAndHost({ files });
  const extractor = new Extractor({
    removeDuplicatedFragments: true,
    scriptSourceHelper: createScriptSourceHelper({ languageService, languageServiceHost }),
    debug: () => {},
  });
  return extractor;
}

describe(MarkdownReporter, () => {
  it('should convert from manifest to markdown content', () => {
    const extractor = createExtractor([
      {
        fileName: '/prj-root/src/main.ts',
        content: `
        import gql from 'graphql-tag';
        const fragment = gql\`
          fragment MyFragment on Query {
            hello
          }
        \`;
        const query = gql\`
          \${fragment}
          query MyQuery {
            ...MyFragment
          }
        \`;
        const mutation = gql\`
          mutation Greeting {
            greeting {
              reply
            }
          }
        \`;
      `,
      },
    ]);
    const manifest = extractor.toManifest(extractor.extract(['/prj-root/src/main.ts'], 'gql'));
    const content = new MarkdownReporter().toMarkdownConntent(manifest, {
      baseDir: '/prj-root',
      outputDir: '/prj-root/dist',
    });
    expect(content).toMatchSnapshot();
  });

  it('should convert from manifest to markdown content with ignoreFragments: false', () => {
    const extractor = createExtractor([
      {
        fileName: '/prj-root/src/main.ts',
        content: `
        import gql from 'graphql-tag';
        const fragment = gql\`
          fragment MyFragment on Query {
            hello
          }
        \`;
        const query = gql\`
          \${fragment}
          query MyQuery {
            ...MyFragment
          }
        \`;
        const mutation = gql\`
          mutation Greeting {
            greeting {
              reply
            }
          }
        \`;
      `,
      },
    ]);
    const manifest = extractor.toManifest(extractor.extract(['/prj-root/src/main.ts'], 'gql'));
    const content = new MarkdownReporter().toMarkdownConntent(manifest, {
      ignoreFragments: false,
      baseDir: '/prj-root',
      outputDir: '/prj-root/dist',
    });
    expect(content).toMatchSnapshot();
  });
});
