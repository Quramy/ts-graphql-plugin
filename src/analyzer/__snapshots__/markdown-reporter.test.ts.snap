// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownReporter should convert from manifest to markdown content 1`] = `
"# Extracted GraphQL Operations
## Queries

### MyQuery

\`\`\`graphql
fragment MyFragment on Query {
  hello
}

query MyQuery {
  ...MyFragment
}
\`\`\`

From [src/main.ts:8:27](../src/main.ts#L8-L13)
    
## Mutations

### Greeting

\`\`\`graphql
mutation Greeting {
  greeting {
    reply
  }
}
\`\`\`

From [src/main.ts:14:30](../src/main.ts#L14-L20)
    
---
Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)"
`;

exports[`MarkdownReporter should convert from manifest to markdown content with ignoreFragments: false 1`] = `
"# Extracted GraphQL Operations
## Queries

### MyQuery

\`\`\`graphql
fragment MyFragment on Query {
  hello
}

query MyQuery {
  ...MyFragment
}
\`\`\`

From [src/main.ts:8:27](../src/main.ts#L8-L13)
    
## Mutations

### Greeting

\`\`\`graphql
mutation Greeting {
  greeting {
    reply
  }
}
\`\`\`

From [src/main.ts:14:30](../src/main.ts#L14-L20)
    
## Fragments

### MyFragment

\`\`\`graphql
fragment MyFragment on Query {
  hello
}
\`\`\`

From [src/main.ts:3:30](../src/main.ts#L3-L7)
    
---
Extracted by [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin)"
`;
